### PR TITLE
feat(react): add useSocialLogin hook

### PR DESCRIPTION
**Pull request by jpklzm**
## Description
This adds a new hook to handle social authentication

## Checklist

- [x] The commit message follows our guidelines
- [x] Tests for the respective changes have been added
- [x] The code is commented, particularly in hard-to-understand areas
- [x] The labels and/or milestones were added

## Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

**Review by paulohfera — APPROVED** [review]
Approved

File: packages/react/src/authentication/hooks/useSocialLogin.ts [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+  type BlackoutError,
+  type Config,
+  postAccountLink,
+  type PostAccountLinkData,
+  postSocialLogin,
+  type PostSocialLoginData,
+} from '@farfetch/blackout-client';
+import { useCallback, useReducer, useRef } from 'react';
+
+const actionTypes = {
```
Comment (ruifcnunes) [inline packages/react/src/authentication/hooks/useSocialLogin.ts:11]: Please separate this logic and place it on the redux package. Make sure to use the correct prefixes on each actionType.

File: packages/react/src/authentication/hooks/useSocialLogin.ts [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+  PostUserSocialLoginRequest: 'POST_USER_SOCIAL_LOGIN_REQUEST',
+  PostUserSocialLoginSuccess: 'POST_USER_SOCIAL_LOGIN_SUCCESS',
+  PostUserSocialLoginFailure: 'POST_USER_SOCIAL_LOGIN_FAILURE',
+  PostAccountLinkRequest: 'POST_ACCOUNT_LINK_REQUEST',
+  PostAccountLinkSuccess: 'POST_ACCOUNT_LINK_SUCCESS',
+  PostAccountLinkFailure: 'POST_ACCOUNT_LINK_FAILURE',
+} as const;
+
+type State = {
```
Comment (ruifcnunes) [inline packages/react/src/authentication/hooks/useSocialLogin.ts:20]: Same as above.

File: packages/react/src/authentication/hooks/useSocialLogin.ts [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+  isLoading: boolean;
+  error: BlackoutError | null;
+  result?: object;
```
Comment (ruifcnunes) [inline packages/react/src/authentication/hooks/useSocialLogin.ts:23]: Can this type be more explicit? do we know what is the structure of the result?

File: packages/react/src/authentication/hooks/useSocialLogin.ts [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+  | PostUserSocialLoginFailureAction
+  | PostUserSocialLoginRequestAction
+  | PostUserSocialLoginSuccessAction
+  | PostAccountLinkFailureAction
+  | PostAccountLinkRequestAction
+  | PostAccountLinkSuccessAction;
+
+const initialState: State = null;
+
+function reducer(state: State, action: Action) {
```
Comment (ruifcnunes) [inline packages/react/src/authentication/hooks/useSocialLogin.ts:71]: Please move this logic to the redux package as well (applies to all redux logic).

File: packages/react/src/authentication/hooks/useSocialLogin.ts [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+  switch (action.type) {
+    case actionTypes.PostUserSocialLoginRequest: {
+      const newState = {
+        isLoading: true,
+        currentRequestId: action.meta.requestId,
```
Comment (ruifcnunes) [inline packages/react/src/authentication/hooks/useSocialLogin.ts:76]: Will we need to manage/store data for more than one request? Should we consider having a `result` property as an object where each result is keyed by its requestId?

something like:
```
result = {
 [action.meta.requestId]: actualResultForTheRequestId
}
```

(same would apply for isLoading and error properties, so we could control them per request).

File: packages/react/src/authentication/hooks/useSocialLogin.ts [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+  };
+};
+
+type PostUserSocialLoginRequestAction = FetchActionBase & {
+  type: typeof actionTypes.PostUserSocialLoginRequest;
+};
+
+type PostUserSocialLoginSuccessAction = FetchActionBase & {
+  type: typeof actionTypes.PostUserSocialLoginSuccess;
+  payload: object;
```
Comment (ruifcnunes) [inline packages/react/src/authentication/hooks/useSocialLogin.ts:39]: Can this type be more explicit?

File: packages/react/src/authentication/hooks/useSocialLogin.ts [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+  payload: BlackoutError;
+};
+
+type PostAccountLinkRequestAction = FetchActionBase & {
+  type: typeof actionTypes.PostAccountLinkRequest;
+};
+
+type PostAccountLinkSuccessAction = FetchActionBase & {
+  type: typeof actionTypes.PostAccountLinkSuccess;
+  payload: object;
```
Comment (ruifcnunes) [inline packages/react/src/authentication/hooks/useSocialLogin.ts:53]: same as above

**Review by ruifcnunes — CHANGES_REQUESTED** [review]

File: packages/react/src/authentication/hooks/__tests__/useSocialLogin.test.tsx [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+  return current;
+};
+
+describe('useSocialLogin', () => {
+  beforeEach(jest.clearAllMocks);
+
+  afterEach(cleanup);
+
+  it('should return correctly with initial state and call all hook dependencies with the correct options', () => {
```
Comment (ruifcnunes) [inline packages/react/src/authentication/hooks/__tests__/useSocialLogin.test.tsx:52]: Can this test description be reviewed, please? 
This hook does not have other hook dependencies (apart from `useAction`) and you're expecting that nothing gets called at this point.

File: packages/react/src/authentication/hooks/__tests__/useSocialLogin.test.tsx [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    expect(current).toStrictEqual(genericMock);
+
+    expect(postSocialLogin).not.toHaveBeenCalled();
+    expect(postAccountLink).not.toHaveBeenCalled();
+  });
+
+  describe('actions', () => {
+    describe('login', () => {
+      it('should call `useSocialLogin` login action without config', async () => {
+        const current = getRenderedHook(mockInitialState);
```
Comment (ruifcnunes) [inline packages/react/src/authentication/hooks/__tests__/useSocialLogin.test.tsx:64]: Can we also have a test where we actually pass in a config for the client call, please?

File: packages/react/src/authentication/hooks/__tests__/useSocialLogin.test.tsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+        await current.actions.login(mockLoginData);
+
+        expect(postSocialLogin).toHaveBeenCalledWith(mockLoginData, undefined);
+      });
+    });
+
+    describe('accountLink', () => {
+      it('should call `useSocialLogin` accountLink action without config', async () => {
```
Comment (ruifcnunes) [inline packages/react/src/authentication/hooks/__tests__/useSocialLogin.test.tsx:73]: same as https://github.com/Farfetch/blackout/pull/837/files#r1311398368

File: packages/react/src/authentication/hooks/useSocialLogin.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+      return await createAccountLink(data, config);
+    },
+    [createAccountLink],
+  );
+
+  return {
+    actions: {
+      login,
+      accountLink,
```
Comment (pedro-gbf) [inline packages/react/src/authentication/hooks/useSocialLogin.ts:41]: I would change name of the function to better indicate the action to be performed, `createAccountLink` is more indicative than simply `accountLink`, just a suggestion if you disagree you can close this thread.

**Review by pedro-gbf — APPROVED** [review]

**Review by ruifcnunes — APPROVED** [review]

**Pull request merged**